Plugin system

mariots opened the pull request:
Solves #8 This PR includes an updated plugin engine architecture and allowing the engine to register callback functions. The callback functions will request certain streams and may be called with the streams as arguments.

mariots committed:
Updated plugin architecture, moving to engine based

mariots committed:
Added file to test engine

mariots committed:
Successfully pass self pluginEngine object to plugin

mariots committed:
Added registering callbacks to engine

mariots committed:
Added dictionaries to hold the streams requested and getters

mariots committed:
Updated architecture, callbacks are held in dictionaries with stream req

mariots committed:
Updated comments and added second stream to test

mariots committed:
Cleaned up files

alexweav commented on plugins/fakePlugin.py:
```diff
@@ -1,12 +1,17 @@
-    def __init__(self):
-        return
+    def init(self):
+        self.pluginEngine.RegisterCallback("realtime", "HelloWorld", self.HelloWorld, ["cameraStills", "otherStream"])
```
we probably don't need this to pass the strings "realtime" and "HelloWorld." we're probably not gonna get to the historical engine (and that would be determined by plugin location anyway), and the HelloWorld string is redundant now that the method is being passed properly. Just function pointer and stream names is enough

mariots replied:
The dictionary needs a string to map to the array that holds the function pointer and the streams. So for this plugin, "HelloWorld" is the key to map to the values [self.HelloWorld, ["cameraStills", "otherStream"]]

alexweav replied:
function pointers are hashable in python, why not just have it be the key?

mariots replied:
Didn't know function pointers are hashable! I'll update it to be the key!

alexweav commented on PluginSystem/PluginEngine.py:
```diff
@@ -8,14 +8,16 @@
 """
 class PluginEngine:
+    historicalCallbacks = {}
+    realtimeCallbacks = {}
```
probably not necessary to have both. historical engine and realtime engine would ideally be entirely different classes, so no need to have both of these in one. we probably won't get to historical stuff at all, so just having one plain PluginEngine with just pluginCallbacks is probably best

alexweav commented on PluginSystem/PluginEngine.py:
```diff
@@ -53,6 +54,29 @@ def LoadFromDirectory(self, folderPath):
+    """
+    def RegisterCallback(self, engine, name, fn, streams):
+
+        if(engine == 'realtime'):
```
no need to make distinction see above

alexweav commented on PluginSystem/PluginEngine.py:
```diff
@@ -53,6 +54,29 @@ def LoadFromDirectory(self, folderPath):
+    """
+    Returns a dictionary of all the historicalCallbacks
+    """
+    def GetHistoricalCallbacks(self):
```
same see above

alexweav commented on PluginSystem/PluginEngine.py:
```diff
@@ -53,6 +54,29 @@ def LoadFromDirectory(self, folderPath):
+    """
+    Returns a dictionary of all the realtimeCallbacks
+    """
+    def GetRealtimeCallbacks(self):
```
GetCallbacks or GetPluginCallbacks or something

alexweav reviewed (changes requested):
good stuff, but since we're low on time and probably not doing historical, theres no need to distinguish between the two for now. if we decide to do historical, it would likely be in a class anyway with its own dedicated plugin folder

mariots committed:
Removed historical and realtime callbacks and replaced with generic

mariots commented:
@alexweav I removed the dual callback functions and replaced them with a generic GetPluginCallbacks. The only thing I didn't change was the user assigning a name to their plugin. We need the name for the dictionary.

mariots committed:
Removed name from registerPluginCallback, replaced with function pointer

mariots committed:
Updated comment in streams
